### PR TITLE
chunk cache

### PR DESCRIPTION
**Pull request by assafvayner**
Chunk Cache implementation.

- caching xorb ranges on file system

### CODE REVIEW

**Comment by assafvayner** [comment]
This PR needs more iterations

**Comment by rajatarya** [comment]
Are you ready for this PR to be reviewed now?

**Comment by assafvayner** [comment]
yes

File: chunk_cache/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+pub use disk_cache::DiskCache;
+
+pub trait ChunkCache {
+    fn get(&mut self, key: &Key, range: &Range) -> Result<Option<Vec<u8>>, ChunkCacheError>;
+    fn put(
+        &mut self,
+        key: &Key,
+        range: &Range,
+        chunk_byte_indicies: &[u32],
```
Comment (rajatarya) [inline chunk_cache/src/lib.rs:15]: typo: indicies -> indices

(in chunk_byte_indicies).

File: chunk_cache/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+mod disk_cache;
+pub mod error;
+
+use cas_types::{Key, Range};
+use error::ChunkCacheError;
+
+pub use disk_cache::DiskCache;
+
+pub trait ChunkCache {
```
Comment (rajatarya) [inline chunk_cache/src/lib.rs:9]: Is the practice to put comments on the trait or on the implementation?

Comment (rajatarya) [reply chunk_cache/src/lib.rs:9]: I'd like to know why this is a trait, what its responsibilities are, and what its intended uses are.

File: chunk_cache/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+#[derive(Debug, Error)]
+pub enum ChunkCacheError {
+    #[error("IO: {0}")]
+    IO(#[from] std::io::Error),
+    #[error("ParseError: {0}")]
+    Parse(String),
+    #[error("bad range")]
+    BadRange,
+    #[error("cache is empty when it is presumed no empty")]
```
Comment (rajatarya) [inline chunk_cache/src/error.rs:13]: Maybe rephrase: "cache is unexpectedly empty"

File: chunk_cache/src/disk_cache.rs [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+
+use crate::{error::ChunkCacheError, ChunkCache};
+
+mod cache_file_header;
+mod file_name;
+
+const BASE64_ENGINE: GeneralPurpose = BASE64_URL_SAFE;
+
+#[derive(Debug, Clone)]
+pub struct DiskCache {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:26]: Please add comments on the pub struct.

What is the purpose of this struct? What should someone looking at this code 3mo from now without any other context understand about it?

File: chunk_cache/src/disk_cache.rs [hunk 5]
```diff
@@ -0,0 +1,574 @@
...
+    cache_root: PathBuf,
+    total_bytes: u64,
+    items: HashMap<(Key, u32, u32), (FileName, u64)>,
+    capacity: Option<u64>,
+}
+
+impl DiskCache {
+    pub fn initialize<T: Into<PathBuf>>(
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:34]: Please add comments on pub fn - especially things like initialize().

Comment (rajatarya) [reply chunk_cache/src/disk_cache.rs:34]: Important to explain the template/generic usage here.

File: chunk_cache/src/disk_cache.rs [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+            items,
+        })
+    }
+
+    fn get_impl(
+        &mut self,
+        key: &Key,
+        range: &Range,
+    ) -> Result<Option<Vec<u8>>, crate::error::ChunkCacheError> {
+        let mut hits = self
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:104]: Please add a comment explaining this code - I believe it is identifying if there are hits in the cache.

File: chunk_cache/src/disk_cache.rs [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+}
+
+impl DiskCache {
+    pub fn initialize<T: Into<PathBuf>>(
+        cache_root: T,
+        capacity: Option<u64>,
+    ) -> Result<Self, ChunkCacheError> {
+        let cache_root = cache_root.into();
+        let mut total_bytes = 0;
+        let mut items = HashMap::new();
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:40]: Is this HashMap thread-safe/concurrency controlled data structures in Rust?

File: chunk_cache/src/disk_cache.rs [hunk 8]
```diff
@@ -0,0 +1,574 @@
...
+        key: &Key,
+        range: &Range,
+    ) -> Result<Option<Vec<u8>>, crate::error::ChunkCacheError> {
+        let mut hits = self
+            .items
+            .iter()
+            .filter(|((k, start, end), _)| *start <= range.start && *end >= range.end && k == key)
+            .peekable();
+        let hit_option = hits.peek();
+        let ((key, start, end), (file_name, len)) = match hit_option {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:110]: What happens if more than one entry in the filter call? This seems to only capture the 1st entry.

File: chunk_cache/src/disk_cache.rs [hunk 9]
```diff
@@ -0,0 +1,574 @@
...
+
+        let result = self.hit(&mut file, *start, range)?;
+
+        Ok(Some(result))
+    }
+
+    fn hit<R: Read + Seek>(
+        &self,
+        file: &mut R,
+        start: u32,
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:143]: How is start different from Range.start?

Is start actually offset into the file?

File: chunk_cache/src/disk_cache.rs [hunk 10]
```diff
@@ -0,0 +1,574 @@
...
+    }
+
+    fn put_impl(
+        &mut self,
+        key: &Key,
+        range: &Range,
+        chunk_byte_indicies: &[u32],
+        data: &[u8],
+    ) -> Result<(), ChunkCacheError> {
+        /* // match for coalescing opportunities
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:174]: Either add more comments explaining why commented out or delete. Commented out code getting checked in just makes maintainability harder later.

File: chunk_cache/src/disk_cache.rs [hunk 11]
```diff
@@ -0,0 +1,574 @@
...
+        let matches = self
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:175]: I suggest renaming this to be more indicative of its intent. This is looking to see if cache already contains this range. Maybe rename to something like, "already_cached" or maybe even better would be to call a private method 'is_cache_hit' and only have the logic for determining if a range is already in the cache in one place.

Right now this is spread across `get_impl` and `put_impl`.

File: chunk_cache/src/disk_cache.rs [hunk 12]
```diff
@@ -0,0 +1,574 @@
...
+            std::fs::remove_file(&file_path)?;
+            return Err(ChunkCacheError::parse("file checksum mismatch"));
+        }
+
+        let result = self.hit(&mut file, *start, range)?;
+
+        Ok(Some(result))
+    }
+
+    fn hit<R: Read + Seek>(
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:140]: I suggest renaming this to read_from_cache or something more clear. Reading 'hit' leads me to think this is determining if something is a hit or not - instead of actually being a read operation.

File: chunk_cache/src/disk_cache.rs [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+            .cloned()
+            .collect::<Vec<_>>();
+        */
+        let matches = self
+            .items
+            .keys()
+            .filter(|(k, start, end)| k == key && *start <= range.start && *end >= range.end)
+            .collect::<Vec<_>>();
+        if !matches.is_empty() {
+            return Ok(());
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:192]: Shouldn't this check that the checksum matches? Maybe I'm reading this too quickly but I believe this just checks for hashmap existence.

File: chunk_cache/src/disk_cache.rs [hunk 14]
```diff
@@ -0,0 +1,574 @@
...
+
+        let hash = blake3::Hasher::new()
+            .update(&header_buf)
+            .update(data)
+            .finalize();
+        let file_name = FileName::new(range.start, range.end, hash);
+        let file_path = Path::join(dir, Into::<PathBuf>::into(&file_name));
+
+        if !dir.exists() {
+            std::fs::create_dir_all(dir)?;
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:210]: Add an `info!()` here indicating cache being created in `dir`.

File: chunk_cache/src/disk_cache.rs [hunk 15]
```diff
@@ -0,0 +1,574 @@
...
+        while capacity <= self.total_bytes && i < 10 {
+            i += 1;
+            self.maybe_evict_one()?;
+        }
+
+        Ok(())
+    }
+
+    // assumes the cache is in the right state for eviction
+    fn maybe_evict_one(&mut self) -> Result<(), ChunkCacheError> {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:249]: This should be renamed to `evict_one`. It isn't choosing to evict - it assumes the cache is already at capacity.

File: chunk_cache/src/disk_cache.rs [hunk 16]
```diff
@@ -0,0 +1,574 @@
...
+
+        if self.total_bytes <= capacity {
+            return Ok(());
+        }
+
+        // evict max 10 items
+        let mut i = 0;
+        while capacity <= self.total_bytes && i < 10 {
+            i += 1;
+            self.maybe_evict_one()?;
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:242]: Without checking for being under capacity - since each cache entry can be of variable size, how do you know you've gotten under capacity with this call?

File: chunk_cache/src/disk_cache.rs [hunk 17]
```diff
@@ -0,0 +1,574 @@
...
+}
+
+fn cache_file_path(cache_root: PathBuf, key: &Key, file_name: &FileName) -> PathBuf {
+    cache_root
+        .join(key_to_subdir(key))
+        .join(Into::<PathBuf>::into(file_name))
+}
+
+impl ChunkCache for DiskCache {
+    fn get(
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:280]: Comments on this implementation. Need to understand the cache behavior in the docs without having to read all the code. Imagine you are looking to integrate this crate 3mo from now without any context - what do you need to know to evaluate whether this ChunkCache is useful for you.

File: chunk_cache/src/disk_cache.rs [hunk 18]
```diff
@@ -0,0 +1,574 @@
...
+
+        Ok(())
+    }
+
+    fn cache_file_path(&self, key: &Key, file_name: &FileName) -> PathBuf {
+        cache_file_path(self.cache_root.clone(), key, file_name)
+    }
+}
+
+fn cache_file_path(cache_root: PathBuf, key: &Key, file_name: &FileName) -> PathBuf {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:273]: Pull into DiskCache impl, no need to be outside it. Not expected to be called from anything except other impl functions.

File: chunk_cache/src/disk_cache.rs [hunk 19]
```diff
@@ -0,0 +1,574 @@
...
+        key: &Key,
+        range: &Range,
+        chunk_byte_indicies: &[u32],
+        data: &[u8],
+    ) -> Result<(), crate::error::ChunkCacheError> {
+        self.put_impl(key, range, chunk_byte_indicies, data)
+    }
+}
+
+fn key_to_subdir(key: &Key) -> String {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:299]: Pull into DiskCache impl, no need to be outside it. Not expected to be called from anything except other impl functions.

File: chunk_cache/src/disk_cache.rs [hunk 20]
```diff
@@ -0,0 +1,574 @@
...
+    let prefix_bytes = key.prefix.as_bytes();
+    let mut buf = vec![0u8; size_of::<MerkleHash>() + prefix_bytes.len()];
+    buf[..size_of::<MerkleHash>()].copy_from_slice(key.hash.as_bytes());
+    buf[size_of::<MerkleHash>()..].copy_from_slice(prefix_bytes);
+    BASE64_ENGINE.encode(buf)
+}
+
+fn subdir_to_key<T: AsRef<[u8]>>(subdir: T) -> Result<Key, ChunkCacheError> {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:307]: Pull into DiskCache impl, no need to be outside it. Not expected to be called from anything except other impl functions.

File: chunk_cache/src/disk_cache.rs [hunk 21]
```diff
@@ -0,0 +1,574 @@
...
+    use std::path::Path;
+
+    use cas_types::{Key, Range};
+    use merklehash::MerkleHash;
+    use rand::Rng;
+
+    pub const DEFAULT_CAPACITY: Option<u64> = Some(16 << 20);
+    pub const RANGE_LEN: u32 = 4000;
+
+    pub fn print_directory_contents(path: &Path) {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache.rs:333]: I think this should be the Display trait implementation for DiskCache.

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+};
+
+use crate::error::ChunkCacheError;
+
+pub struct CacheFileHeader {
+    pub chunk_byte_indicies: Vec<u32>,
+    pub header_len: usize,
+}
+
+impl CacheFileHeader {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:13]: Comments needed here.

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 23]
```diff
@@ -0,0 +1,55 @@
+use std::{
+    io::{Read, Seek, Write},
+    mem::size_of,
+};
+
+use crate::error::ChunkCacheError;
+
+pub struct CacheFileHeader {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:8]: Comments needed here.

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+    pub chunk_byte_indicies: Vec<u32>,
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:9]: Typo: indicies -> indices

Comment (rajatarya) [reply chunk_cache/src/disk_cache/cache_file_header.rs:9]: Also, docstring, what is this vector capturing?

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+    pub header_len: usize,
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:10]: Docstring, Is this the length of this struct?

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+impl CacheFileHeader {
+    pub fn new<T: Into<Vec<u32>>>(chunk_byte_indicies: T) -> Self {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:14]: Comments

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+        let chunk_byte_indicies = chunk_byte_indicies.into();
+        let header_len = (chunk_byte_indicies.len() + 1) * size_of::<u32>();
+        Self {
+            chunk_byte_indicies,
+            header_len,
+        }
+    }
+
+    pub fn deserialize<R: Read + Seek>(reader: &mut R) -> Result<Self, ChunkCacheError> {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:23]: Comments.

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+            let idx = read_u32(reader)?;
+            chunk_byte_indicies.push(idx);
+        }
+        Ok(Self {
+            chunk_byte_indicies,
+            header_len: (chunk_byte_indicies_len as usize + 1) * size_of::<u32>(),
+        })
+    }
+
+    pub fn serialize<W: Write>(&self, writer: &mut W) -> Result<usize, std::io::Error> {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:37]: Comments

File: chunk_cache/src/disk_cache/cache_file_header.rs [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+        let mut num_written = write_u32(writer, self.chunk_byte_indicies.len() as u32)?;
+        for idx in &self.chunk_byte_indicies {
+            num_written += write_u32(writer, *idx)?;
+        }
+        Ok(num_written)
+    }
+}
+
+pub fn read_u32<R: Read>(reader: &mut R) -> Result<u32, std::io::Error> {
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/cache_file_header.rs:46]: Why are `read_u32` and `write_u32` outside of the impl? They seem like useful helper functions, but they should be in the impl, right?

File: chunk_cache/src/disk_cache/file_name.rs [hunk 30]
```diff
@@ -0,0 +1,87 @@
+use std::{mem::size_of, path::PathBuf};
+
+use base64::{engine::GeneralPurpose, prelude::BASE64_URL_SAFE, Engine};
+use blake3::Hash;
+
+use crate::error::ChunkCacheError;
+
+const BASE64_ENGINE: GeneralPurpose = BASE64_URL_SAFE;
+/// A file name is represented as the start index and end index of chunks for the given xorb
+/// and a timestamp of last successful access or put
```
Comment (rajatarya) [inline chunk_cache/src/disk_cache/file_name.rs:10]: Is timestamp missing from the struct? Or should this part be removed from comment?

**Comment by rajatarya** [comment]
Does the file IO handle partially written files? Do you want to leverage atomic file moves in *nix OSes? (write cache file to tmp dir and then move it to final location).
